packages/core/src/FrameLoop.ts

```diff
@@ -159,50 +172,53 @@ export class FrameLoop {
         const decay = config.decay === true ? 0.998 : config.decay
         position =
           from +
-          (velocity / (1 - decay)) *
-            (1 - Math.exp(-(1 - decay) * (time - startTime)))
+          (v0 / (1 - decay)) *
+            (1 - Math.exp(-(1 - decay) * animated.elapsedTime!))
+
+        // derivative of position
+        velocity = v0 * Math.exp(-(1 - decay) * animated.elapsedTime!)
 
         finished = Math.abs(animated.lastPosition - position) < 0.1
         if (finished) to = position
       }
       // Spring easing
       else {
-        let lastTime = animated.lastTime !== void 0 ? animated.lastTime : time
-        if (animated.lastVelocity !== void 0) {
-          velocity = animated.lastVelocity
-        }
+        velocity =
+          animated.lastVelocity !== void 0 ? animated.lastVelocity : v0 * 1000
+
+        const freq = Math.sqrt(config.tension! / config.mass!) * (2 * Math.PI)
```

Comment:
Why is `* (2 * Math.PI)` needed?

Reply:
Ok, so I asked help around to sort this out and just pushed new code:

```js
const w0 = Math.sqrt(config.tension! / config.mass!)
const dt = 100 / w0
const numSteps = Math.ceil(deltaTime / dt)

for (let n = 0; n < numSteps; ++n) {
  const springForce = -config.tension! * (position - to)
  const dampingForce = -config.friction! * (velocity * 1000)
  const acceleration = (springForce + dampingForce) / config.mass!
  velocity = velocity + (acceleration * dt) / (1000 * 1000) // expressed in ms
  position = position + velocity * dt
}
```

**Disclaimer: the following is only my understanding**

I think I was confused by the term natural frequency, which isn't what `w0` is. The correct term for `w0` is angular frequency.

The stability criterion from the [article, equation 23 page 23](https://hplgit.github.io/num-methods-for-PDEs/doc/pub/vib/pdf/vib-4print-A4-2up.pdf) @mtiller was referring to is valid only for undamped harmonic oscillator. I suspect that the right formula for damped harmonic oscillator is actually `dt < Math.sqrt(2) / w0` instead of `dt < 2 / w0`.

The stability criterion makes sure your oscillator is stable (meaning it will get to its destination eventually), but there will be a margin of error, and the closer you get to the stability limit, the more the spring will behave erratically compared to the analytical solution.

Here, I've set `dt = 0.1 / w` (actually `dt = 100/w` as these are `ms`), and it seems to behave pretty close to when `dt` was `1ms`. My observation is that we move from **10-20 steps** to just **2-3 steps** in average for the following config: `config: { tension: 120, friction: 12, mass: 1 }`.